org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/BoundSet.java

```diff
@@ -871,20 +872,19 @@ private boolean isParameterizedDependency(TypeBound typeBound) {
 	}
 
 	private ConstraintTypeFormula incorporateIntoParameterizedDependencyIfAllArgumentsAreProperTypes(TypeBound typeBound,
-			TypeBound[] firstBounds, TypeBound[] nextBounds) {
-		Collection<TypeBound> properTypesForAllInferenceVariables = getProperTypesForAllInferenceVariablesOrNull((ParameterizedTypeBinding)typeBound.right, firstBounds, nextBounds);
+			TypeBound[] firstBounds, TypeBound[] nextBounds, Map<InferenceVariable,TypeBound> properTypesByInferenceVariable) {
+		if(properTypesByInferenceVariable.isEmpty()) {
+			return null;
+		}
+		Collection<TypeBound> properTypesForAllInferenceVariables = getProperTypesForAllInferenceVariablesOrNull((ParameterizedTypeBinding)typeBound.right, firstBounds, nextBounds, properTypesByInferenceVariable);
 		if (null != properTypesForAllInferenceVariables) {
 			return combineWithProperTypes(properTypesForAllInferenceVariables, typeBound);
 		}
 		return null;
 	}
 
 	private Collection<TypeBound> getProperTypesForAllInferenceVariablesOrNull(ParameterizedTypeBinding parameterizedType,
-			TypeBound[] firstBounds, TypeBound[] nextBounds) {
-		final Map<InferenceVariable,TypeBound> properTypesByInferenceVariable = properTypesByInferenceVariable(firstBounds, nextBounds);
-		if(properTypesByInferenceVariable.size() == 0) {
-			return null;
-		}
+			TypeBound[] firstBounds, TypeBound[] nextBounds, Map<InferenceVariable,TypeBound> properTypesByInferenceVariable) {
```

Comment:
`firstBounds` and `nextBounds` are no longer used in this method. Removal should propagate to callers several levels out (thus reverting their introduction in 372c6048)

Reply:
good. done.